### PR TITLE
Bunstore accounts operations fix

### PR DESCRIPTION
**Pull request by nicholaspcr**
#### Summary

References #5816
Does mainly changes to operations related to the account on `delete` and `purge` in the `User` and `Organization` stores.

#### Changes

- Add `WithSoftDelete` wrapper on bunstore methods which are responsible for deleting user related info. 
- Add account operations on user and organization delete/purge operations.

#### Testing

Manual testing:
- create 3 users and 3 organizations
  - Delete usr-2
  - Purge usr-3
  - Purge usr-2
- Same for orgs.

Looking at DB at each step and seeing if it matches the expected result.

##### Regressions

This is a fix for a bug, should not change anything besides the necessary steps to delete/purge accounts with their respective user/organization.

#### Notes for Reviewers

I'm not closing the issue with this PR since I want to do some tests regarding that, but was hoping to do some changes like  putting the account operations around an interface and include its validations on the tests. Plan to add that while working on https://github.com/TheThingsIndustries/lorawan-stack/issues/3426#event-7357538269

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [ ] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: pkg/identityserver/bunstore/api_key_store.go [hunk 1]
```diff
@@ -359,7 +359,7 @@ func (s *apiKeyStore) DeleteEntityAPIKeys(ctx context.Context, entityID *ttnpb.E
 	))
 	defer span.End()
 
-	entityType, entityUUID, err := s.getEntity(ctx, entityID)
+	entityType, entityUUID, err := s.getEntity(store.WithSoftDeleted(ctx, false), entityID)
```
Comment (adriansmares) [inline pkg/identityserver/bunstore/api_key_store.go:362]: Was this possible with `gorm`, but not `bun` ?

File: pkg/identityserver/bunstore/organization_store.go [hunk 2]
```diff
@@ -510,6 +510,24 @@ func (s *organizationStore) DeleteOrganization(ctx context.Context, id *ttnpb.Or
 		return wrapDriverError(err)
 	}
 
+	accountModel, err := s.getAccountModel(ctx, id.GetEntityIdentifiers().EntityType(), id.GetOrganizationId())
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return store.ErrUserNotFound.WithAttributes(
+				"user_id", id.GetOrganizationId(),
```
Comment (adriansmares) [inline pkg/identityserver/bunstore/organization_store.go]: `organization_id`.

File: pkg/identityserver/bunstore/organization_store.go [hunk 3]
```diff
@@ -543,6 +561,26 @@ func (s *organizationStore) RestoreOrganization(ctx context.Context, id *ttnpb.O
...
```
Comment (adriansmares) [inline pkg/identityserver/bunstore/organization_store.go]: `organization_id`.

File: pkg/identityserver/user_registry.go [hunk 4]
```diff
@@ -693,12 +693,13 @@ func (is *IdentityServer) deleteUser(ctx context.Context, ids *ttnpb.UserIdentif
 		return nil, err
 	}
 	err := is.store.Transact(ctx, func(ctx context.Context, st store.Store) error {
-		err := st.DeleteUser(ctx, ids)
+		// Delete the the user's sessions to enforce logouts.
```
Comment (adriansmares) [inline pkg/identityserver/user_registry.go:696]: I guess this triggered your `WithSoftDelete` changes, right ?

**Review by adriansmares — COMMENTED** [review]
Can you please add the `CreateAfterPurge` tests to the entities please ?

File: pkg/identityserver/bunstore/api_key_store.go (hunk 1 above)
Comment (nicholaspcr) [reply pkg/identityserver/bunstore/api_key_store.go:362]: https://github.com/TheThingsNetwork/lorawan-stack/blob/ac05bfe303c360973eae155d3b827bfa7daa8f50/pkg/identityserver/gormstore/api_key_store.go#L221-L226

But I did not check for everyone, just changed in a way that local test would work. Going to double check it

File: pkg/identityserver/user_registry.go (hunk 4 above)
Comment (nicholaspcr) [reply pkg/identityserver/user_registry.go:696]: Not really, just ended up changing the order and therefore updating the comments.

The reason for changing the order was to keep it the same logic as the purge, where you try to delete all related things before deleting said user.

File: pkg/identityserver/storetest/user_store.go [hunk 5]
```diff
@@ -373,6 +373,58 @@ func (st *StoreTest) TestUserStoreCRUD(t *T) {
...
 	})
+
+	t.Run("CreateAfterPurge", func(t *T) {
+		a, ctx := test.New(t)
+		var err error
+		start := time.Now().Truncate(time.Second)
+		stamp := start.Add(-1 * time.Minute)
+
+		// UserId have to be the same as the one deleted in the previous test case.
+		created, err = s.CreateUser(ctx, &ttnpb.User{
```
Comment (adriansmares) [inline pkg/identityserver/storetest/user_store.go]: We already test `Create` above, so feel free to provide only `Ids` and `PrimaryEmailAddress` to have a minimal test. No need to assert what is returned, only the lack of errors.

I also think we should do `Create` -> `Delete` -> `Restore` -> `Get` -> `Delete` -> `Purge`, 2 times in a loop.

**Review by adriansmares — APPROVED** [review]

**Review by kschiffer — APPROVED** [review]

**Pull request merged**